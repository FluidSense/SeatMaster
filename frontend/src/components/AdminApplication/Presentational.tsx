import React from 'react';
import { IPostAdminApplicationForm } from '../../API/interfaces';
import { IApplication } from '../Application';
import ApplicationOverview from '../ApplicationReview/ApplicationOverview';
import AssignSeat from '../AssignSeat';
import { IRoom } from '../ViewRooms';
import AcceptApplication from './AcceptApplication';
import './adminapplication.css';
import ApplicationSeatDisplay from './ApplicationSeatDisplay';
import SetWaitingList from './SetWaitingList';

interface IProps {
  rooms: IRoom[];
  application: IApplication;
  removeStudentFromSeat: (roomId: number, seatId: string) => void;
  updateApplication: (id: number, app: IPostAdminApplicationForm) => void;
}

const Presentational: React.FunctionComponent<IProps> = (props) => {
  const { application, rooms, removeStudentFromSeat, updateApplication } = props;
  if (!(application && rooms)) return null;
  const givenSeat = application.seat;
  const givenRoomId = givenSeat ? givenSeat.roomId : 0;
  const selectedRooms = rooms.filter(obj => obj.id === givenRoomId);

  return (
    <div className="main-content">
      <ApplicationOverview
        application={application}
        title={application.user ? application.user.fullname : ''}
        pathToEdit={`/admin/applications/${application.id}/edit`}
      />
      <div className="edit-lower-row">
        <AssignSeat rooms={rooms} application={application} />
        <ApplicationSeatDisplay
          seat={application.seat}
          room={selectedRooms[0]}
          removeFromSeat={removeStudentFromSeat}
        />
      </div>
      <div className="center-div">
        <SetWaitingList
          seat={application.seat}
          updateApplication={updateApplication}
          application={application}
        />
        <AcceptApplication application={application} updateApplication={updateApplication}/>
      </div>
    </div>
  );
};

export default Presentational;
