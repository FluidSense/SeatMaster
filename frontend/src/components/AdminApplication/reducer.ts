import { AnyAction } from 'redux';
import { IApplication } from '../Application';
import {
  FETCH_APPLICATION_DIRECTLY,
  FETCH_APPLICATION_DIRECTLY_FAILED,
  RESET_REVIEW_APPLICATION_STATUS_CODE,
} from './constants';

export interface IAdminReviewApplicationState {
  application?: IApplication;
  api: {
    status: number,
  };
}

const initialState: IAdminReviewApplicationState = {
  api: {
    status: 0,
  },
  application: undefined,
};

export const AdminReviewApplicationReducer = (
  state: IAdminReviewApplicationState = initialState,
  action: AnyAction,
): IAdminReviewApplicationState => {
  switch (action.type) {
    case FETCH_APPLICATION_DIRECTLY: {
      return {
        ...state,
        api: {
          status: 200,
        },
        application: action.payload,
      };
    }
    case FETCH_APPLICATION_DIRECTLY_FAILED: {
      return {
        ...state,
        api: {
          status: 404,
        },
        application: initialState.application,
      };
    }
    case RESET_REVIEW_APPLICATION_STATUS_CODE: {
      return {
        ...state,
        api: initialState.api,
      };
    }
    default:
      return { ...state };
  }
};
