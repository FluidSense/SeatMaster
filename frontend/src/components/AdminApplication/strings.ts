export const _NAME = 'Name';
export const _EMAIL = 'E-Mail';
export const _PHONE = 'Phone';
export const _PARTNER = 'Partner';
export const _PREFERRED_ROOM = 'Preferred room';
export const _SEAT_ROLLOVER = 'Seat Rollover';
export const _NEEDS = 'Needs';
export const _EDIT_APPLICATION = 'Edit Application';
export const _MASTER_STATUS = 'Master status';
export const _CHANGE_STUDENT_WARNING = 'A student already has a seat here. '
  + 'Do you want to change their allocation?';
export const _REMOVE_STUDENT_FROM_SEAT = 'Remove student from seat';
export const _REMOVE_FROM_SEAT = 'Remove from seat';
export const _APPROVE_APP = 'Approve application';
export const _REVOKE_APP = 'Revoke approval';