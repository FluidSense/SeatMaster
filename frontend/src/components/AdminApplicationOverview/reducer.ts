import { AnyAction } from 'redux';
import { IApplication } from '../Application';
import { APP_NOT_FOUND } from '../commonConstants';
import {
  SUCCESSFULL_APPLICATION_UPDATE,
  UNSUCCESSFULL_APPLICATION_UPDATE,
} from '../EditApplication/constants';
import { GET_ALL_APPLICATIONS } from './constants';

export interface IApplicationState {
  applications: IApplication[];
  registeredApplication: IApplication;
  api: {
    status: number,
  };
}

const initialState: IApplicationState = {
  api: {
    status: 0,
  },
  applications: [],
  registeredApplication: {
    status: APP_NOT_FOUND,
  },
};

export const ApplicationReducer = (
  state: IApplicationState = initialState,
  action: AnyAction,
): IApplicationState => {
  switch (action.type) {
    case GET_ALL_APPLICATIONS:
      const applications = action.payload;
      return {
        ...state,
        applications,
      };
    case 'SET_APPLICATION_DATA': {
      return {
        ...state,
        registeredApplication: action.payload,
      };
    }
    case 'REMOVE_APPLICATION_DATA': {
      return {
        ...state,
        registeredApplication: initialState.registeredApplication,
      };
    }
    case SUCCESSFULL_APPLICATION_UPDATE: {
      return {
        ...state,
        api: {
          ...state.api,
          status: 200,
        },
      };
    }
    case UNSUCCESSFULL_APPLICATION_UPDATE: {
      return {
        ...state,
        api: {
          ...state.api,
          status: 400,
        },
      };
    }
    default:
      return { ...state };
  }
};
