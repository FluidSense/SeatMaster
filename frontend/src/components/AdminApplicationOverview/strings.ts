export const _LINK_NAME_TITLE = 'Name';
export const _LINK_APP_STATUS = 'Application Status';
export const _LINK_MASTER_STATUS = 'Master Status';
export const _APPLICATIONS_TITLE = 'Applications';
