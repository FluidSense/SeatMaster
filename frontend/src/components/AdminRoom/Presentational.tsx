import AlertStripe from 'nav-frontend-alertstriper';
import HovedKnapp from 'nav-frontend-knapper';
import { Input, Textarea } from 'nav-frontend-skjema';
import { Sidetittel } from 'nav-frontend-typografi';
import React, { ChangeEvent, SyntheticEvent } from 'react';
import { ETIKETT_WARNING } from '../commonConstants';
import {
  _ALERT_CREATED_MESSAGE,
  _BUTTON_CREATE_ROOM,
  _BUTTON_DELETE_ROOM,
  _BUTTON_UPDATE_ROOM,
  _INPUT_LABEL_NAME,
  _INPUT_LABEL_NOTES,
  _TITLE_CREATE_NEW_ROOM,
  _TITLE_UPDATE_NEW_ROOM,
} from './strings';

interface IProps {
  roomName: string;
  roomNotes: string;
  setNotes: (roomNotes: ChangeEvent<HTMLInputElement>) => void;
  setName: (roomName: ChangeEvent<HTMLInputElement>) => void;
  buttonDisabled: boolean;
  onClick: () => void;
  deleteRoom: () => void;
  showAlert: boolean;
  alertMessage?: string;
  roomExists: boolean;
}

const Presentational: React.FunctionComponent<IProps> = (props) => {
  const {
    roomName,
    roomNotes,
    roomExists,
    setNotes,
    setName,
    buttonDisabled,
    onClick,
    deleteRoom,
    showAlert,
    alertMessage,
  } = props;
  const displayAlert =
    showAlert
      ? (
        <AlertStripe solid={true} type={ETIKETT_WARNING}>
          {alertMessage}
        </AlertStripe>)
      : null;
  const titleText = roomExists ? _TITLE_UPDATE_NEW_ROOM : _TITLE_CREATE_NEW_ROOM;
  const buttonText = roomExists ? _BUTTON_UPDATE_ROOM : _BUTTON_CREATE_ROOM;
  const deleteButton =
    roomExists
      ? (
        <HovedKnapp id={'delete-room-button'} type="fare" onClick={deleteRoom}>
          {_BUTTON_DELETE_ROOM}
        </HovedKnapp>)
      : null;
  // TextArea returns the wrong type, so its type has to be forced
  const assertEventType = (event: SyntheticEvent<EventTarget, Event>) => {
    const changeEvent = event as ChangeEvent<HTMLInputElement>;
    setNotes(changeEvent);
  };

  return (
    <>
      <Sidetittel>{titleText}</Sidetittel>
      {displayAlert}
      <Input
        id={'input-room-name'}
        onChange={setName}
        value={roomName}
        label={_INPUT_LABEL_NAME}
        bredde={'XL'}
      />
      <Textarea
        id={'input-room-notes'}
        onChange={assertEventType}
        value={roomNotes}
        label={_INPUT_LABEL_NOTES}
      />
      <div id="state-buttons">
        <HovedKnapp
          id={'create-room-button'}
          type="hoved"
          disabled={buttonDisabled}
          onClick={onClick}
        >
          {buttonText}
        </HovedKnapp>
        {deleteButton}
      </div>
    </>
  );
};

export default Presentational;
