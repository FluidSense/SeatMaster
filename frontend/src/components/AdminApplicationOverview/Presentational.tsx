import { Sidetittel } from 'nav-frontend-typografi';
import React, { FunctionComponent } from 'react';
import { IApplication } from '../Application';
import { TitleAndSpinner } from '../LoadingPageSpinner/TitleAndSpinner';
import SearchBar, { searchBarEvent } from '../SearchBar';
import { IRoom } from '../ViewRooms';
import ApplicationLink from './ApplicationLink';
import ApproveSeated from './ApproveSeated';
import { FETCHING_APPLICATION_DATA } from './constants';
import PutOnWaitingList from './PutOnWaitingList';
import { _APPLICATIONS_TITLE } from './strings';

interface IProps {
  applications: IApplication[];
  rooms: IRoom[];
  fetching: string;
  approve: (ids: number[]) => void;
  filterFunction: (event: searchBarEvent) => void;
  putWaiting: (ids: number[]) => void;
}
const title = <Sidetittel>{_APPLICATIONS_TITLE}</Sidetittel>;
const emptyPage = (filterFunction: (event: searchBarEvent) => void) => (
  <div className="main-content">
    {title}
    <SearchBar filterFunction={filterFunction}/>
</div>);
const emptyPageWithLoader = (<TitleAndSpinner title={_APPLICATIONS_TITLE}/>);

const Presentational: FunctionComponent<IProps> = (props) => {
  const { applications, rooms, fetching, filterFunction, approve, putWaiting } = props;
  if (fetching === FETCHING_APPLICATION_DATA) return emptyPageWithLoader;
  if (!applications || !applications.length) return emptyPage(filterFunction);
  const applicationList = applications.map(application => (
    <ApplicationLink key={application.id} application={application} rooms={rooms} />
    ));
  return (
    <div className="main-content">
      <div className="title-and-button">
        <Sidetittel>
          {_APPLICATIONS_TITLE}
        </Sidetittel>
        <PutOnWaitingList putWaiting={putWaiting} applications={applications}/>
        <ApproveSeated approve={approve} applications={applications}/>
      </div>
      <SearchBar filterFunction={filterFunction}/>
      {applicationList}
    </div>
  );
};

export default Presentational;
