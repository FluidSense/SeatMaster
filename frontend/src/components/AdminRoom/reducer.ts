import { AnyAction } from 'redux';
import { CREATE_ROOM, DELETE_ROOM, UPDATE_ROOM } from './constants';
import { _ALERT_CREATED_MESSAGE, _ALERT_DELETED_MESSAGE, _ALERT_UPDATED_MESSAGE } from './strings';

export interface IState {
  error?: string;
  submitted: boolean;
}

const initialState = {
  submitted: false,
};

const reducer = (
  state: IState = initialState,
  action: AnyAction,
): IState => {
  const { type, payload } = action;
  switch (type) {
    case CREATE_ROOM:
      if (payload) return { ...state, submitted: payload };
      return { ...state, error: _ALERT_CREATED_MESSAGE };
    default:
      return state;
  }
};

export default reducer;
