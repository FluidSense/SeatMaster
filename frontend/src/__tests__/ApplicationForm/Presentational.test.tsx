import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import fetchMock from 'fetch-mock';
import AlertStripe from 'nav-frontend-alertstriper';
import KnappBase from 'nav-frontend-knapper';
import * as React from 'react';
import Presentational from '../../components/ApplicationForm/Presentational';
import { _ALERT_USER_ERROR } from '../../components/ApplicationForm/strings';
import { IRegisteredUserState } from '../../components/RegisterUser/reducer';

jest.mock('../../store', () => ({
  getState: jest.fn(() => ({ oidc: { user: { id_token: 'test' } } })),
}));

describe('application form', () => {
  const changeModalMock = () => null;
  const POST_FORM_DATA = 'http://localhost:5000/application/';
  const user = {
    email: 'london@starbucks.com',
    fullname: 'Starbucks London',
    phone: '22225555',
    status: '',
    username: 'Sutaba',
  };
  const userInfo: IRegisteredUserState = {
    admin: false,
    loading: false,
    registered: true,
  };
  const doNothing = () => { return; };
  it('renders alertBox correctly', () => {
    const wrapper = shallow(
      <Presentational
        setApplication={changeModalMock}
        modalIsOpen={true}
        userInformation={userInfo}
        changeModal={changeModalMock}
        rooms={[]}
        getRooms={doNothing}
      />);
    const preClickAlert = wrapper.find(AlertStripe);
    expect(preClickAlert.length).toBe(0);
    wrapper.setState({ error: _ALERT_USER_ERROR });
    const alert = wrapper.find(AlertStripe);
    expect(alert.length).toBe(1);
  });

  it('renders correctly', () => {
    const wrapper = shallow(
      <Presentational
        setApplication={changeModalMock}
        modalIsOpen={false}
        userInformation={userInfo}
        changeModal={changeModalMock}
        rooms={[]}
        getRooms={doNothing}
      />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
