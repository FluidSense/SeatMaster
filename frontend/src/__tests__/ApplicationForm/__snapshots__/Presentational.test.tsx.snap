// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`application form renders correctly 1`] = `
<Fragment>
  <form
    id="new-application-form"
    onSubmit={[Function]}
  >
    <p
      style={
        Object {
          "fontStyle": "italic",
        }
      }
    >
      Additional information about your courses at NTNU will be gathered for the purpose of finding your master status.
    </p>
    <ApplicationFormPersonal
      isAdmin={false}
    />
    <ApplicationFormPreferences
      hasPartner={false}
      isAdmin={false}
      rooms={Array []}
      updateApplicationFormData={[Function]}
    />
    <ApplicationFormComments
      hasNeeds={false}
      isAdmin={false}
      updateApplicationFormData={[Function]}
    />
    <KnappBase
      autoDisableVedSpinner={true}
      className={null}
      disabled={false}
      htmlType="submit"
      id="submit-application"
      inaktivKlasseVedDisabled={false}
      mini={false}
      spinner={false}
      type="hoved"
    >
      Submit
    </KnappBase>
  </form>
</Fragment>
`;
